### PR TITLE
MdeModulePkg: Add alternative queue sizes in NVME driver

### PR DESCRIPTION
**Pull request by VivianNK**
# Description

Add a new PCD that toggles between the old driver behavior and new behavior that uses a different hardcoded queue size to support specific devices.  Default behavior should not change.

MdeModulePkg/NvmExpressDxe: Improve NVMe controller init robustness

It has been observed that some faulty NVMe devices may return invalid values. The code in NvmExpressDxe recognizes the controller is not responding correctly and issues an ASSERT() often in DEBUG builds or a reset in RELEASE builds.

The following changes are made to NvmeControllerInit() to prevent a faulty NVMe device from halting the overall boot:

1. Check the Vendor ID and Device ID to verify they are read properly and if not, return EFI_DEVICE_ERROR
2. After the ASSERT() when Memory Page Size Minimum (Cap.Mpsmin),
   produce an error message and return EFI_DEVICE_ERROR

MdeModulePkg/NvmExpressDxe: Correct function parameter modifer

Updates the `Cap` parameter for `ReadNvmeControllerCapabilities()` to be `OUT` since the buffer pointed to is written within the function.

---
Cherry-picked the following commits:
https://github.com/microsoft/mu_basecore/commit/9837db39da115a1fa4a322d13adafbba862212d8
https://github.com/microsoft/mu_basecore/commit/0a5a1d950449f1133a6a648efe849e4be5cbb287
https://github.com/microsoft/mu_basecore/commit/e397168aec57fa86c3683d97022111d76daf9fd6

---

- [ ] Breaking change?
- [ ] Impacts security?
- [ ] Includes tests?

## How This Was Tested

Tested in Project MU

## Integration Instructions

N/A

### CODE REVIEW

File: MdeModulePkg/MdeModulePkg.dec [hunk 1]
```diff
@@ -1187,6 +1187,9 @@
   # @Prompt Delay access XHCI register after it issues HCRST (us)
   gEfiMdeModulePkgTokenSpaceGuid.PcdDelayXhciHCReset|2000|UINT16|0x30001060
 
+  # Support alternative hardware queue sizes in NVME driver
```
Comment (mdkinney) [inline MdeModulePkg/MdeModulePkg.dec:1190]: Add @Prompt and description of TRUE and FALSE setting and the default value.

Also, is this PCD really required?  What would be the impact if just assuming this is always enabled.  Prefer to not add new PCDs unless absolutely required.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c [hunk 2]
```diff
@@ -730,13 +730,34 @@ NvmeControllerInit (
...
+  UINT16               VidDid[2];
   UINT8                Sn[21];
   UINT8                Mn[41];
 
+  PciIo = Private->PciIo;
+
+  //
+  // Verify the controller is still accessible
+  //
+  Status = PciIo->Pci.Read (
```
Comment (mdkinney) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:742]: Looks unrelated to new features/PCD.  Should it be split into a different commit?

Comment (VivianNK) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:742]: Returned old commits, unsquashed.
